### PR TITLE
fix(state-sync): Correct StateRoot when the last chunk is missing

### PR DESCRIPTION
**Pull request by nikurt**
Missing a chunk leads to incorrect state root in state sync.

### CODE REVIEW

**Comment by nikurt** [comment]
cc: @ppca

File: integration-tests/src/tests/nearcore/sync_state_nodes.rs [hunk 1]
```diff
@@ -530,3 +535,141 @@ fn sync_state_dump() {
...
+
+                let tx = SignedTransaction::send_money(
+                    i + 1,
+                    "test0".parse().unwrap(),
+                    "test1".parse().unwrap(),
+                    &signer,
+                    1,
+                    *genesis_block.hash(),
+                );
+                assert_eq!(env.clients[0].process_tx(tx, false, false), ProcessTxResponse::ValidTx);
```
Comment (ppca) [inline integration-tests/src/tests/nearcore/sync_state_nodes.rs:633]: I think doing `env.clients[0].process_tx()` here means this transaction will actually be included in the next block not this one, is this intended?

File: integration-tests/src/tests/nearcore/sync_state_nodes.rs [hunk 2]
```diff
@@ -530,3 +535,141 @@ fn sync_state_dump() {
...
+
+            let genesis_block = env.clients[0].chain.get_block_by_height(0).unwrap();
+            let mut blocks = vec![genesis_block.clone()];
+            let signer =
+                InMemorySigner::from_seed("test0".parse().unwrap(), KeyType::ED25519, "test0");
+            let target_height = epoch_length + 1;
+            for i in 1..=target_height {
+                let block = env.clients[0].produce_block(i).unwrap().unwrap();
+                blocks.push(block.clone());
+                if (i % epoch_length) != 0
```
Comment (ppca) [inline integration-tests/src/tests/nearcore/sync_state_nodes.rs:602]: doesn't this mean the last block of an epoch won't be missing any chunks? I think the last block of an epoch missing chunks is exactly what we want to test for?

File: integration-tests/src/tests/nearcore/sync_state_nodes.rs [hunk 3]
```diff
@@ -530,3 +535,141 @@ fn sync_state_dump() {
...
+                    *genesis_block.hash(),
+                );
+                assert_eq!(env.clients[0].process_tx(tx, false, false), ProcessTxResponse::ValidTx);
+            }
+
+            // Simulate state sync
+
+            // No blocks were skipped, therefore we can compute the block height of the first block of the current epoch.
+            let sync_hash_height = ((target_height / epoch_length) * epoch_length + 1) as usize;
+            let sync_hash = *blocks[sync_hash_height].hash();
```
Comment (ppca) [inline integration-tests/src/tests/nearcore/sync_state_nodes.rs:640]: the index here should be `sync_hash_height -1`?

File: integration-tests/src/tests/nearcore/sync_state_nodes.rs [hunk 4]
```diff
@@ -530,3 +535,141 @@ fn sync_state_dump() {
...
+                        0,
+                        &state_root,
+                        PartId::new(i, num_parts),
+                        &state_parts[i as usize],
+                        &epoch_id,
+                    )
+                    .unwrap();
+            }
+        }
+    });
```
Comment (ppca) [inline integration-tests/src/tests/nearcore/sync_state_nodes.rs:674]: do we want to also do `set_state_header` and `set_state_finalize` for clients[1] and also check that the data of clients[1] is the same as data of clients[0] after state sync? like `test_catchup_gas_price_change()`

**Comment by nikurt** [comment]
The logic of state sync finalization is fine. It doesn't need me fixing it.
Finalization works as following:
* Find the last chunk of an epoch
* Include it in state_sync_header
* Reprocess that chunk
Doesn't matter at which height that chunk was included.

My test fails at the creation of state parts, because of step 2) in the following case:
1) Imagine that the last chunk of a shard was a while ago. More than the finality. For instance, 5 blocks ago state_root was X, and 4 blocks ago it became Y.
2) `obtain_state_part()` expect flat storage to contain state corresponding to X.
3) When another node finalizes the sync, it will take state dump corresponding to X and replay that chunk from 5 blocks before the end of the epoch.

The problem is in step 2). Flat Storage keeps updating its head every block for every shard, regardless if that shard had chunks included at a height or not.
The problem is magnified by state snapshots moving flat head to the sync_prev_prev_hash. But even without that, if several chunks in a raw were missing, state X won't be recoverable from flat storage.

cc: @Longarithm 
cc: @ppca

**Pull request closed without merging**